### PR TITLE
Add `partial_transpose` to `quantum_info.linalg_operations`

### PR DESCRIPTION
**Pull request by renatomello**
Needs to be merged before #1456 

Checklist:
- [x] Reviewers confirm new code works as expected.
- [x] Tests are passing.
- [x] Coverage does not decrease.
- [x] Documentation is updated.

### CODE REVIEW

**Review by BrunoLiegiBastonLiegi — COMMENTED** [review]
I think it might be useful to allow the inputs to both `partial_transpose` and `partial_trace` (this one in a separate PR though) to be batches of states, similarly to what it is done for `vectorization` in #1459. This would make more efficient, for instance, the calculation of the `meyer_wallach_entanglement` 
https://github.com/qiboteam/qibo/blob/165b2046b169c70426c780b7035403bba10807e3/src/qibo/quantum_info/entanglement.py#L196
and similar methods that need to perform these operations on several different states.

**Comment by renatomello** [comment]
> I think it might be useful to allow the inputs to both `partial_transpose` and `partial_trace` (this one in a separate PR though) to be batches of states, similarly to what it is done for `vectorization` in #1459. This would make more efficient, for instance, the calculation of the `meyer_wallach_entanglement`
> 
> https://github.com/qiboteam/qibo/blob/165b2046b169c70426c780b7035403bba10807e3/src/qibo/quantum_info/entanglement.py#L196
> 
> and similar methods that need to perform these operations on several different states.

Done.

File: tests/test_quantum_info_operations.py [hunk 1]
```diff
@@ -113,6 +114,98 @@ def test_partial_trace(backend, density_matrix):
...
+
+        # partial transpose of two-qubit werner state is known analytically
+        target = (1 / 4) * np.array(
+            [
+                [1 - p, 0, 0, -2 * p],
+                [0, p + 1, 0, 0],
+                [0, 0, p + 1, 0],
+                [-2 * p, 0, 0, 1 - p],
+            ],
+            dtype=complex,
```
Comment (BrunoLiegiBastonLiegi) [inline tests/test_quantum_info_operations.py:195]:
```suggestion
        state, target = _werner_state(p, backend)
```
is this supposed to be the `_werner_state` above?

File: tests/test_quantum_info_operations.py [hunk 2]
```diff
@@ -113,6 +114,98 @@ def test_partial_trace(backend, density_matrix):
...
+        )
+        target = backend.cast(target, dtype=target.dtype)
+
+        transposed = partial_transpose(state, [1], backend)
+
+        if batch:
+            print(transposed.shape)
```
Comment (BrunoLiegiBastonLiegi) [inline tests/test_quantum_info_operations.py:202]:
```suggestion
```

**Comment by renatomello** [comment]
@AlejandroSopena @stavros11 this is ready. Could you review it soon since it is blocking #1456 ?

**Review by stavros11 — APPROVED** [review]
Thanks @renatomello. Generally fine with me, but I have not checked validity so I just trust the test.

It took me some time to understand the supported input shapes and I think we are kind of overcomplicating it by supporting 1-dimensional arrays. To me, since this is a partial trace, it would even be sufficient to stick to 2-dimensional. If someone wants to process a 1-dimensional one, they can still do the outer product before calling this (this way it is also a bit more modular). However, I am not the best candidate user of this, so it's not a strong opinion.

Other than that, I understand you are also supporting a "batch" dimension (`N`). There may be a way to generalize this to arbitrary number of batch dimensions (`(..., k, k)`), so that you don't even have to distinguish between batch/no batch, but I haven't thought much about whether this can be easily done and maybe it's not very useful in the end.

**Comment by renatomello** [comment]
> Thanks @renatomello. Generally fine with me, but I have not checked validity so I just trust the test.
> 
> It took me some time to understand the supported input shapes and I think we are kind of overcomplicating it by supporting 1-dimensional arrays. To me, since this is a partial trace, it would even be sufficient to stick to 2-dimensional. If someone wants to process a 1-dimensional one, they can still do the outer product before calling this (this way it is also a bit more modular). However, I am not the best candidate user of this, so it's not a strong opinion.

I think it's fine supporting statevectors because it takes an extra step from the users' hands at the price of a handful of lines of code on our side. So the trade-off is worth IMO.
 
> Other than that, I understand you are also supporting a "batch" dimension (`N`). There may be a way to generalize this to arbitrary number of batch dimensions (`(..., k, k)`), so that you don't even have to distinguish between batch/no batch, but I haven't thought much about whether this can be easily done and maybe it's not very useful in the end.

I like the idea, but for the interest of time I prefer to move forward as is.

**Review by Simone-Bordoni — APPROVED** [review]

**Pull request merged**